src/test/java/com/uber/cadence/workflow/WorkflowTest.java

```diff
@@ -1470,6 +1570,75 @@ public void testAsyncRetry() {
     assertTrue(trace.get(2).startsWith("retry at "));
   }
 
+  public static class TestAsyncRetryOptionsChangeWorkflow implements TestWorkflow2 {
+
+    private final List<String> trace = new ArrayList<>();
+
+    @Override
+    public String execute(boolean useExternalService) {
+      RetryOptions retryOptions;
+      if (Workflow.isReplaying()) {
+        retryOptions =
+            new RetryOptions.Builder()
+                .setMinimumAttempts(1)
+                .setMaximumInterval(Duration.ofSeconds(1))
+                .setInitialInterval(Duration.ofSeconds(1))
+                .setMaximumAttempts(3)
+                .build();
+      } else {
+        retryOptions =
+            new RetryOptions.Builder()
+                .setMinimumAttempts(2)
+                .setMaximumInterval(Duration.ofSeconds(1))
+                .setInitialInterval(Duration.ofSeconds(1))
+                .setMaximumAttempts(2)
+                .build();
+      }
+
+      trace.clear(); // clear because of replay
+      trace.add("started");
+      Async.retry(
+              retryOptions,
+              () -> {
+                trace.add("retry at " + Workflow.currentTimeMillis());
+                return Workflow.newFailedPromise(new IllegalThreadStateException("simulated"));
+              })
+          .get();
+      trace.add("beforeSleep");
+      Workflow.sleep(60000);
+      trace.add("done");
+      return "";
+    }
+
+    @Override
+    public List<String> getTrace() {
+      return trace;
+    }
+  }
+
+  /** @see DeterministicRunnerTest#testRetry() */
+  @Test
+  public void testAsyncRetryOptionsChange() {
```

Comment:
Both of this and the sync test exercises the WorkflowRetryerInternal.retryAsync path. Should add a test for WorkflowRetryerInternal.retry as well.